libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt

```diff
@@ -397,23 +400,23 @@ internal class MapboxTripSession(
         navigatorJobController.scope.launch {
             val currentDate = Date()
             navigator.updateLocation(rawLocation, currentDate)
-            updateDataFromNavigatorStatus(currentDate)
+            updateDataFromNavigatorStatus(time.nanoTime())
         }
 
         unconditionalStatusPollingJob = navigatorJobController.scope.launch {
             delay(UNCONDITIONAL_STATUS_POLLING_PATIENCE)
             while (isActive) {
                 launch {
-                    updateDataFromNavigatorStatus(Date())
+                    updateDataFromNavigatorStatus(time.nanoTime())
```

Comment:
Which clock `Time.nanoTime` eventually uses? We need to make sure that it uses [`SystemClock.elapsedRealtimeNanos`](https://developer.android.com/reference/android/os/SystemClock#elapsedRealtimeNanos()) - it is the only clock that can be reliably compared to values returned from [`Location.getElapsedRealtimeNanos`](https://developer.android.com/reference/android/location/Location#getElapsedRealtimeNanos())

Reply:
And also - I couldn't find the place where we craft `FixLocation` from Android's `Location`. What value do we set to `monotonicTimestampNanoseconds`?

Reply:
@SiarheiFedartsou The get status calls use it's [System.nanoTime()](https://github.com/mapbox/mapbox-navigation-android/blob/384a25d5e1a409d6316fe3a1dfd9f4c9747ea81a/libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/Time.kt#L8). This is the equivalent of SystemClock.elapsedRealtimeNanos, SystemClock is a Android interface where System.nanoTime is Java, but the underlying clock is the same.

Also here is the Location->FixLocation mapping, and it is using Location.elapsedRealtimeNanos
https://github.com/mapbox/mapbox-navigation-android/pull/3451/files#diff-97b9704c96e95f49d4e77e9d4ac685e1R40

🕥  🕥  🕥

Reply:
@SiarheiFedartsou 
> Which clock `Time.nanoTime` eventually uses? We need to make sure that it uses [`SystemClock.elapsedRealtimeNanos`](https://developer.android.com/reference/android/os/SystemClock#elapsedRealtimeNanos()) - it is the only clock that can be reliably compared to values returned from [`Location.getElapsedRealtimeNanos`](https://developer.android.com/reference/android/location/Location#getElapsedRealtimeNanos())

Does that mean that NN requires [SystemClock.elapsedRealtime](https://developer.android.com/reference/android/os/SystemClock#elapsedRealtime()) instead of [System.nanoTime](https://www.tutorialspoint.com/java/lang/system_nanotime.htm)? cc @RingerJK @kmadsen 

> And also - I couldn't find the place where we craft `FixLocation` from Android's `Location`. What value do we set to `monotonicTimestampNanoseconds`?

https://github.com/mapbox/mapbox-navigation-android/blob/351030e4780fec72820257958e97c54a78ac844e/libnavigator/src/main/java/com/mapbox/navigation/navigator/LocationEx.kt#L13

https://github.com/mapbox/mapbox-navigation-android/blob/351030e4780fec72820257958e97c54a78ac844e/libnavigator/src/main/java/com/mapbox/navigation/navigator/LocationEx.kt#L40

Reply:
@kmadsen beat me to it 💪

Reply:
@SiarheiFedartsou https://github.com/mapbox/mapbox-navigation-android/blob/master/libnavigator/src/main/java/com/mapbox/navigation/navigator/LocationEx.kt
Where we cast Location to FixLocation

Reply:
As @kmadsen mentioned it shouldn't be any difference 👍

Reply:
`time.nanoTime` is actually `System.nanoTime()` https://github.com/mapbox/mapbox-navigation-android/blob/master/libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/Time.kt

Reply:
@kmadsen Is it documented somewhere that System.nanoTime and SystemClock.elapsedRealtimeNanos will have the same source?

Reply:
> @ kmadsen Is it documented somewhere that System.nanoTime and SystemClock.elapsedRealtimeNanos will have the same source?

It's a fair question, and I have been assuming that interfaces use the same interfaces into a single counting mechanism. Mobile clocks have become a slightly hot topic with the importance of GPS accuracy, so maybe this will change, but I think we can continue to assume a single clock until we see support for multiple. I don't know of documentation that supports these assumptions.

But we can look at java native source to double check
[native System.nanoTime](https://android.googlesource.com/platform/libcore/+/7757924/luni/src/main/native/java_lang_System.cpp#98)
[native SystemClock.elapsedRealtimeNanos](https://android.googlesource.com/platform/system/core/+/master/libutils/SystemClock.cpp#68)

both are using `clock_gettime`

Reply:
Okay, but why don't we just use `SystemClock.elapsedRealtimeNanos` as recommended in Android docs(see https://developer.android.com/reference/android/location/Location#getElapsedRealtimeNanos())? It seems to be trivial to use it instead `System.nanoTime`. Also nobody guarantee that implementation of any of `System.nanotTime`/`SystemClock.elapsedRealtimeNanos` won't change  in the future?
I do not insist, but just want to understand the reason why we have to use `System.nanoTime` instead of recommended `SystemClock.elapsedRealtimeNanos`?

Reply:
The results are the same regardless of the interface. The argument for `System.nanoTime` is to prefer Java only and cross platform interfaces. The argument for `SystemClock.elapsedRealtimeNanos` is as you stated, use Android specific interfaces.

It's nice to use these time values interchangeably, because some libraries are Java only. We're also interested in computing elapsed time with navigation-native interfaces, and it has access to [steady_clock](https://en.cppreference.com/w/cpp/chrono/steady_clock). @SiarheiFedartsou can you do a test and compare steady_clock with System.nanoTime?

> "Also nobody guarantees that implementation of any of System.nanoTime/SystemClock.elapsedRealtimeNanos won't change in the future."

We also have not provided proof that they will ever diverge. For what it's worth, I've been using both of them for years and all systems are up.

Reply:
> We're also interested in computing elapsed time with navigation-native interfaces, and it has access to steady_clock. @SiarheiFedartsou can you do a test and compare steady_clock with System.nanoTime?

I think it is not a good idea too. Nobody guarantee that sources of `steady_clock` and `System.nanoTime` will be the same. Even if we will experimentally prove it on some environment.


Just interesting finding on [SO](https://stackoverflow.com/questions/26266441/android-time-intervals-with-deep-sleep-system-nanotime-system-currenttimemi):
```
System.nanoTime() - Works great if Android is running, but stops on deep sleep (this is bad).
System.currentTimeMillis() - Good, but not appropriate because it can be changed by the user or by code with setCurrentTimeMillis(long).
SystemClock.elapsedRealtimeNanos() - Counts elapsed time even on deep sleep, but requires API level 17.
```
If it is true they are not the same.

Reply:
Actually they are different:
```

*/
--
  | int64_t elapsedRealtimeNano()
  | {
  | #if defined(__linux__)
  | struct timespec ts;
  | int err = clock_gettime(CLOCK_BOOTTIME, &ts);
```

```

static jlong System_nanoTime(JNIEnv*, jclass) {
--
  | timespec now;
  | #ifdef CLOCK_MONOTONIC_RAW
  | clock_gettime(CLOCK_MONOTONIC_RAW, &now);
```

Do you see this `CLOCK_BOOTTIME` and `CLOCK_MONOTONIC_RAW`? 
Let's go to the [docs](https://linux.die.net/man/2/clock_gettime):
```
CLOCK_MONOTONIC

Clock that cannot be set and represents monotonic time since some unspecified starting point. This clock is not affected by discontinuous jumps in the system time (e.g., if the system administrator manually changes the clock), but is affected by the incremental adjustments performed by adjtime(3) and NTP.
...
CLOCK_MONOTONIC_RAW (since Linux 2.6.28; Linux-specific)
Similar to CLOCK_MONOTONIC, but provides access to a raw hardware-based time that is not subject to NTP adjustments or the incremental adjustments performed by adjtime(3).
CLOCK_BOOTTIME (since Linux 2.6.39; Linux-specific)
Identical to CLOCK_MONOTONIC, except it also includes any time that the system is suspended. This allows applications to get a suspend-aware monotonic clock without having to deal with the complications of CLOCK_REALTIME, which may have discontinuities if the time is changed using settimeofday(2).
```

It votes for what was written in that SO question.

Reply:
But have you verified use cases where they are actually different? In practice I have not identified a system which actually creates different values.

Reply:
> here, we'd be able to reproduce error when going into deep sleep. Is this a use case? I think it may be, if a system ever goes into deep sleep it would make these two values uncomparable

Exactly. 
What I want to do is to have this really critical part of SDK to be 100% aligned with official Android docs(btw misalignments with them was already a source of problem which we are trying to solve here). If the problem is that `SystemClock` is not cross platform(btw why do we need it to be cross platform?), we can implement two versions of `Time` interface: Android one with `SystemClock.elapsedRealtimeNanoes()` and another one with `System.nanoTime`.

Reply:
Fixing in https://github.com/mapbox/mapbox-navigation-android/pull/3483

Reply:
BTW interesting fact. It seems that `steady_clock` is implemented in the same way as `System.nanoTime`, i.e. it won't increase if system is in deep sleep: https://android.googlesource.com/platform/ndk.git/+/43255f3d58b03cd931d29d1ee4e5144e86e875ce/sources/cxx-stl/llvm-libc++/libcxx/src/chrono.cpp#124

It means we cannot reliably use it too.

Reply:
Thanks for persisting and digging into the details @SiarheiFedartsou 

>  If the problem is that SystemClock is not cross platform(btw why do we need it to be cross platform?

Am still hoping for an interface where we can calculate time between native and the SDK. [For example](https://github.com/mapbox/mapbox-navigation-native/blob/a7e5b8bfd803ee324873e0cd1ec16ef034379e3d/src/mblc/navigator/extrapolator.cpp#L22), can nav-native use an existing api to calculate time estimations?

Or do we need to create an interface where we implement the DeviceClock inside the SDK?

Reply:
There are java only libraries that we use to create time estimates. If we're using a java only library, and we compare with System.nanoTime, it's just worth knowing.

RxJava is a popular library in java. It uses [currentTimeMillis](https://github.com/ReactiveX/RxJava/search?q=currentTimeMillis&unscoped_q=currentTimeMillis) and [System.nanoTime](https://github.com/ReactiveX/RxJava/search?q=System.nanoTime&unscoped_q=System.nanoTime)

Reply:
@kmadsen 
currently we do not request any timestamps from the system(with neither `std::chrono::steady_clock::now()` nor `SystemClock. elapsedRealtimeNanos()` nor other apis). All calculations are performed on timestamps provided in `FixLocation`. 
But in the future we may need it. In this case the best solution would be provide access to platform clock through some interface, but we don't need it right now.

Reply:
> But in the future we may need it. In this case the best solution would be provide access to platform clock through some interface, but we don't need it right now.

+1